Refactor LST inputs

ariostas opened the pull request:
This PR improves a bit the way inputs are passed into LST. Portable collections are constructed from the very beginning, and the code used in CMSSW and standalone was unified. There are some points to discuss that I'll point out in review comments.

ariostas commented on RecoTracker/LSTCore/interface/LSTInput.h:
```diff
@@ -0,0 +1,251 @@
+    std::memcpy(pixelHits.deltaPhi(), deltaPhi_vec.data(), nPixelHits * sizeof(float));
+
+    int pixelSegmentsSize = ptIn_vec.size();
+    // if (pixelSegmentsSize > n_max_pixel_segments_per_module) {
+    // lstWarning(
+    //     "\
+    //     *********************************************************\n\
+    //     * Warning: Pixel line segments will be truncated.       *\n\
+    //     * You need to increase n_max_pixel_segments_per_module. *\n\
+    //     *********************************************************");
+    //   pixelSegmentsSize = n_max_pixel_segments_per_module;
+    // }
```
This is one thing I still need to fix. Do we want to continue imposing some limit, or should we make more of there sizes be dynamic?

ariostas commented on RecoTracker/LSTCore/src/alpaka/LSTEvent.dev.cc:
```diff
@@ -138,82 +108,28 @@ void LSTEvent::addHitToEvent(std::vector<float> const& x,
-    size = n_max_pixel_segments_per_module;
-  }
+void LSTEvent::addPixelSegmentToEventStart(PixelSegmentsHostCollection const* pixelSegmentsHC) {
+  // We can't use CopyToDevice because the device can be DevHost
+  pixelSegmentsDC_.emplace(pixelSegmentsHC->const_view().metadata().size(), queue_);
+  alpaka::memcpy(queue_, pixelSegmentsDC_->buffer(), pixelSegmentsHC->buffer());
```
I think it would be ideal to let the framework take care of the copying. However, I think we can only get constant references to objects, but we need to modify things further. The simplest solution was to just make the copy to device here, but maybe a cleaner solution would be to separate the data that comes from inputs from the one generated in kernels by putting them in separate collections.

ariostas replied:
This is fixed now

ariostas commented:
It's not ready yet. It compiles, but it doesn't run. I'll finish is on Monday, but you can start taking a look if you want.

ariostas commented:
Standalone works now, but I think I'll need some help from @VourMa to get CMSSW to work. I'm currently getting an error saying that there is no dictionary for the new LSTInput collection. I'm not sure what it is, but my guess is that it's probably something with the fillDescriptions method of LSTInputProducer.

VourMa commented:
> Standalone works now, but I think I'll need some help from @VourMa to get CMSSW to work. I'm currently getting an error saying that there is no dictionary for the new LSTInput collection. I'm not sure what it is, but my guess is that it's probably something with the fillDescriptions method of LSTInputProducer.

I can take a look later today/tomorrow. I will message you for the details.

ariostas commented:
In the meantime, here's a timing comparison on cgpu-1:

```
This PR [0bb4f24]
Average time for map loading = 582.717 ms
Average time for input loading = 7165.74 ms
Average time for lst::Event creation = 0.000442778 ms
   Evt    Hits       MD       LS      T3       T5       pLS       pT5      pT3      TC       Reset    Event     Short             Rate
   avg      2.3      0.4      0.4      1.6      1.3      0.6      1.0      0.8      1.2      0.0       9.5       6.6+/-  1.4       9.6   explicit[s=1]
   avg      2.7      0.8      0.7      2.1      1.8      0.9      1.6      1.0      1.8      0.0      13.4       9.9+/-  2.2       6.8   explicit[s=2]
   avg      3.6      1.3      1.4      3.1      2.6      1.1      2.8      1.7      3.4      0.0      21.1      16.3+/-  3.2       5.4   explicit[s=4]
   avg      4.7      2.2      2.2      4.3      3.8      1.5      3.9      2.2      4.8      0.1      29.7      23.4+/-  4.7       5.1   explicit[s=6]
   avg      6.2      3.4      3.0      5.6      5.3      1.9      5.2      2.8      6.3      0.0      39.7      31.6+/-  6.9       5.1   explicit[s=8]

Master [38eca30]
Average time for map loading = 589.504 ms
Average time for input loading = 7417.81 ms
Average time for lst::Event creation = 0.000488191 ms
   Evt    Hits       MD       LS      T3       T5       pLS       pT5      pT3      TC       Reset    Event     Short             Rate
   avg      2.3      0.4      0.4      1.6      1.3      0.6      1.0      0.8      1.2      0.0       9.6       6.6+/-  1.4       9.6   explicit[s=1]
   avg      1.7      0.7      0.7      2.0      1.6      0.8      1.4      0.9      1.9      0.0      11.7       9.2+/-  1.9       5.9   explicit[s=2]
   avg      2.8      1.3      1.4      3.1      2.7      1.1      2.7      1.4      3.3      0.0      19.9      16.0+/-  3.2       5.1   explicit[s=4]
   avg      4.1      2.1      2.2      4.3      3.9      1.5      3.7      2.0      4.9      0.0      28.9      23.3+/-  4.0       4.9   explicit[s=6]
   avg      5.3      3.2      3.1      5.7      5.6      2.0      5.2      2.6      6.5      0.0      39.3      32.0+/-  6.2       5.0   explicit[s=8]
```

The slowdown on the hits column for multiple streams is probably from this part. Any suggestions for a way to fix it? The options I had in mind were just creating an extra vector with references to the right queue, or less preferably, making the queue a public attribute.
https://github.com/SegmentLinking/cmssw/blob/0bb4f24862a5cd6f76b923387166e5931d6e5bdb/RecoTracker/LSTCore/standalone/bin/lst.cc#L408-L413

ariostas commented:
I added the class defs, but I'm still getting the same error. I'm not sure what I'm missing

VourMa commented:
> I added the class defs, but I'm still getting the same error. I'm not sure what I'm missing

I noticed that you added the class definition in the `LST` package. Don't you define the class in the `LSTCore` package though?

slava77 commented on RecoTracker/LSTCore/interface/LSTInputSoA.h:
```diff
@@ -0,0 +1,53 @@
+
+namespace lst {
+
+  GENERATE_SOA_LAYOUT(InputHitsSoALayout,
```
I'm not sure if `HitsBase*` and `HitsExtended*` or similar is a better naming to indicate the content rather than the origin (same for pLS).

Then, the question is, if it's possible to transparently have a non-owning `Hits` SoA view which would directly give access to a union of members in both. Perhaps it's just coding convenience, not an actual performance gain. So, maybe not that essential.

ariostas replied:
I thought about this, but I'm not sure if there's a better way other than to just create some new struct that keeps both views. I'll also ask this in the upstream PR.

ariostas commented on RecoTracker/LSTCore/interface/LSTInput.h:
```diff
@@ -0,0 +1,242 @@
+    charge_vec.reserve(n_see);
+    seedIdx_vec.reserve(n_see);
+    deltaPhi_vec.reserve(n_see);
+    trkX = ph2_x;
```
Thanks! I'll do two `memcpy`s

ariostas commented on RecoTracker/LSTCore/src/alpaka/Hit.h:
```diff
@@ -43,17 +43,18 @@ namespace ALPAKA_ACCELERATOR_NAMESPACE::lst {
     {
       auto geoMapDetId = endcapGeometry.geoMapDetId();  // DetId's from endcap map
       auto geoMapPhi = endcapGeometry.geoMapPhi();      // Phi values from endcap map
+      unsigned int nHits = hits.metadata().size();
```
Asserting that it matches with which other number? I made this change so that it uses the real size of the SoA instead of trusting some input number

slava77 replied:
the code below reads from `inputHits`, but `hits.size` is checked, assuming they are the same size

ariostas replied:
Oh yeah, I'll add an assert

ariostas commented on RecoTracker/LSTCore/src/alpaka/LST.cc:
```diff
@@ -383,8 +150,8 @@ void LST::run(Queue& queue,
 
-  event.addPixelSegmentToEventFinalize(
-      in_hitIndices_vec0_, in_hitIndices_vec1_, in_hitIndices_vec2_, in_hitIndices_vec3_, in_deltaPhi_vec_);
+  // event.addPixelSegmentToEventFinalize(
+  //     in_hitIndices_vec0_, in_hitIndices_vec1_, in_hitIndices_vec2_, in_hitIndices_vec3_, in_deltaPhi_vec_);
```
Sorry, I forgot to remove that

ariostas commented:
I resolved some of the review comments. I'll follow up on the rest. For now, I'll make sure that at least standalone is working.

/run standalone

VourMa commented on RecoTracker/LST/plugins/LSTInputProducer.cc:
```diff
@@ -149,23 +183,29 @@ void LSTPixelSeedInputProducer::produce(edm::StreamID iID, edm::Event& iEvent, c
+                                      ptCut_);
+
+  iEvent.emplace(lstInputPutToken_, std::move(lstInputHC));
   iEvent.emplace(lstPixelSeedsPutToken_, std::move(see_seeds));
```
I don't remember clearly but it's highly probable.
Also, I am unsure how we use these on the device. If they are used for anything, would references from the host work?

VourMa commented on RecoTracker/LSTCore/interface/alpaka/LSTInputDeviceCollection.h:
```diff
@@ -0,0 +1,13 @@
+
+#include "DataFormats/Portable/interface/alpaka/PortableCollection.h"
+
+#include "RecoTracker/LSTCore/interface/alpaka/Common.h"
```
Why is this needed?

ariostas replied:
Just for convenience. Otherwise we'd have to use `lst::` prefixes since in that header we specify that `ALPAKA_ACCELERATOR_NAMESPACE::lst` inherits everything from `::lst`.

VourMa commented on RecoTracker/LSTCore/src/alpaka/LSTEvent.dev.cc:
```diff
@@ -1352,32 +1278,51 @@ int LSTEvent::getNumberOfT5TrackCandidates() {
   } else {
-    if (!hitsHC_) {
+    if (!lstInputHC_) {
       if (inCMSSW) {
-        auto hits_d = hitsDC_->view<HitsSoA>();
-        auto nHits = hits_d.metadata().size();
-        std::array<int, 2> const hits_sizes{{static_cast<int>(nHits), static_cast<int>(nModules_)}};
-        hitsHC_.emplace(hits_sizes, queue_);
-        auto hits_h = hitsHC_->view<HitsSoA>();
+        auto hits_d = lstInputDC_->view<InputHitsSoA>();
+        int nHits = hits_d.metadata().size();
+        std::array<int, 3> const hits_sizes{{nHits, 0, 0}};
+        lstInputHC_.emplace(hits_sizes, queue_);
+        auto hits_h = lstInputHC_->view<InputHitsSoA>();
```
Maybe I am lost in the the way that the code changes appear but it seems to me that in the specific case of LST running on the GPU in standalone mode this function returns only the input hits (hence the function name is misleading), while in other cases it returns the full input? If so, is this on purpose?

ariostas replied:
When running on GPU and `inCMSSW==true` we return only the inputs hits and otherwise the full inputs. I also don't really like this since it seems like an easy way to get confused, but this is what we were already doing (in the interest of saving a bit of time in CMSSW, which arguably doesn't really matter that much).

VourMa replied:
I think that the time to copy everything was not negligible, so I remember/understand now. I guess my issue is that this function now changed name and, even though it is accurate for what happens in CMSSW, in standalone mode it does more than the name implies. Maybe add a comment to clarify?

ariostas replied:
Or what if we make a separate function that returns the trimmed-down data, so it is more evident?

VourMa replied:
That could be a way to go.

ariostas replied:
I separated it into a new function. Could you take a look at it when you get a chance?

VourMa commented on RecoTracker/LST/plugins/alpaka/LSTProducer.cc:
```diff
@@ -89,8 +66,7 @@ namespace ALPAKA_ACCELERATOR_NAMESPACE {
   private:
-    edm::EDGetTokenT<LSTPixelSeedInput> lstPixelSeedInputToken_;
-    edm::EDGetTokenT<LSTPhase2OTHitsInput> lstPhase2OTHitsInputToken_;
+    edm::EDGetTokenT<lst::LSTInputDeviceCollection> lstInputToken_;
```
My SoAs are a bit rusty, so could you remind me at which point we get the `LSTInputDeviceCollection` from the `LSTInputHostCollection`? Because the output of `LSTInputProducer` is `LSTInputHostCollection`, while here we directly ask for `LSTInputDeviceCollection` from the framework. Is it that the framework automatically copies the SoAs to device with the proper type, once we request a `*HostCollection` with a `*DeviceCollection` type?

ariostas replied:
That's how I'm expecting it to work, but I haven't been able to test it yet. At least that's how it works for ESProducers. We put in `LSTESData<DevHost>` and request `LSTESData<Device>` and the framework does the copying.

VourMa reviewed (commented):
I took a look at the code and made some comments, mostly for my understanding.

Does the class issue remain in CMSSW? If so, let me know and I could make some time to run the code myself and try to figure out the problem.

ariostas commented on RecoTracker/LST/plugins/LSTInputProducer.cc:
```diff
@@ -149,23 +183,29 @@ void LSTPixelSeedInputProducer::produce(edm::StreamID iID, edm::Event& iEvent, c
+                                      ph2_x,
+                                      ph2_y,
+                                      ph2_z,
+                                      ptCut_);
```
The function takes references, and (after addressing Slava's comment) there is no copying being done apart from filling the SoAs, so I think it's fine. But I can change it if you prefer.

ariostas commented on RecoTracker/LSTCore/interface/HitsSoA.h:
```diff
@@ -9,12 +9,7 @@
 namespace lst {
 
   GENERATE_SOA_LAYOUT(HitsSoALayout,
```
Yeah, that's right. And the thing is that what we get from CMSSW is constant, so we can't just include the extra columns from the beginning since we wouldn't be able to modify them.

ariostas commented:
> I took a look at the code and made some comments, mostly for my understanding.
>
> Does the class issue remain in CMSSW? If so, let me know and I could make some time to run the code myself and try to figure out the problem.

Thank you for your comments, @VourMa! I replied to them. I'm still having issues with classes, but I'm giving it another try. In the meeting later today I'll let you know how it goes.

ariostas commented:
I think standalone should be fine now.

/run standalone

slava77 commented:
maybe it's easier to submit this as a draft upstream and get feedback there?
I'm not sure the last exchange in https://mattermost.web.cern.ch/cms-o-and-c/pl/jo9f68nzwbnmmkh38dr3qtf3fa was productive

ariostas commented:
> maybe it's easier to submit this as a draft upstream and get feedback there?

Sounds good, I'll rename things as you suggested and make sure that the rest is fine, and then I'll submit it. However, I'll be on Holiday until Tuesday, so I'll do so when I get back

ariostas commented:
Thank you, @VourMa. Since you're the one who knows what needs to change you should go ahead (now or after the other issues are resolved)

VourMa commented:
> Thank you, @VourMa. Since you're the one who knows what needs to change you should go ahead (now or after the other issues are resolved)

Sure! I will do it once the issues are resolved, so that I can properly test it.

ariostas commented:
@slava77 do you think it would be useful to run the tests upstream so they can see the errors in the CI logs? If so, could you start them?

ariostas commented:
@VourMa I think this is ready for you to fix the HLT part.

We still have to decide on the naming and handling of the SoAs that were split. @slava77 should I rename `Input*`, `*` to `*Base`, `*Extended`? And should I define a small struct that stores pointers to these two?

slava77 commented:
maybe it's easier to submit this as a draft upstream and get feedback there?
I'm not sure the last exchange in https://mattermost.web.cern.ch/cms-o-and-c/pl/jo9f68nzwbnmmkh38dr3qtf3fa was productive


> should I rename `Input*`, `*` to `*Base`, `*Extended`? 

I didn't come up with a better suggestion since my earlier review.

> And should I define a small struct that stores pointers to these two?

I don't have a strong preference for this: without writing extra methods that would encapsulate this union, it will not be particularly practical; but to encapsulate may need too much coding.

VourMa commented:
> @VourMa I think this is ready for you to fix the HLT part.

Thanks! Let's discuss in the meeting and I can prepare this for early next week (I would like to see also the bot results in CMSSW).

slava77 commented on RecoTracker/LST/plugins/BuildFile.xml:
```diff
@@ -35,6 +35,8 @@
   <use name="RecoTracker/LST"/>
   <use name="RecoTracker/LSTCore"/>
   <use name="RecoTracker/Record"/>
+  <use name="DataFormats/L1TrackTrigger"/>
+  <use name="Validation/RecoTrack"/>
```
how did this end up here?

slava77 replied:
if `Validation/RecoTrack/interface/trackFromSeedFitFailed.h` is actually needed, it would be better to move it to `DataFormats/TrackReco`

ariostas replied:
Why is it bad to use `Validation/RecoTrack`?

slava77 replied:
just from a common sense of logical dependency (Validation packages contain code that analyzes data created in Reco or more common packages; so this is at least a potential for a circular dependency here).

My comment was also about `DataFormats/L1TrackTrigger`: what triggered this dependency?

ariostas replied:
Ah okay, that makes sense.

`Validation/RecoTrack/interface/trackFromSeedFitFailed.h` was used in `LSTPixelSeedInputProducer.cc` while building the seed collection.
https://github.com/SegmentLinking/cmssw/blob/1c5d583ce8c6cf95d0d140922215d634739ab5b3/RecoTracker/LST/plugins/LSTPixelSeedInputProducer.cc#L112
And we already had the dependency for the host-only plugins
https://github.com/SegmentLinking/cmssw/blob/1c5d583ce8c6cf95d0d140922215d634739ab5b3/RecoTracker/LST/plugins/BuildFile.xml#L23


As for `DataFormats/L1TrackTrigger`, I think I was debugging something and I forgot to remove it.

ariostas commented:
On Manos's branch even standalone was segfaulting, so there must have been some aditional issue. I rebased to the current master and fixed an inconsistency. Standalone seems to be fine, so the issue is probably from the input producer. I'll run the CI to see if the CMSSW plots look any different.

/run all

ariostas commented:
The error was because I needed to adjust the CI.

@VourMa the HLT setup that you had in the other PR works too. (Or at least it runs without crashing, but I haven't looked at the plots)

VourMa commented:
> @VourMa the HLT setup that you had in the other PR works too. (Or at least it runs without crashing, but I haven't looked at the plots)

Thanks! @slava77 Should we rerun the tests on cmssw PR to check also for differences in results?

VourMa commented:
> Thanks! @slava77 Should we rerun the tests on cmssw PR to check for differences in results?

It seems like 0.755 shows differences in the trigger results (even though I don't fully get the output). I think we need to run the HLT plots to make sure what's going on.

VourMa commented:
>  I think we need to run the HLT plots to make sure what's going on.

I tried running in `CMSSW_15_1_X_2025-04-10-2300` (not the latest IB but I am not sure it is connected) and after
`scram b clean; scram b -j 12`
and then running HLT with `alpaka,trackingLST`, I get:
```
----- Begin Fatal Exception 19-Apr-2025 01:38:57 PDT-----------------------
An exception of category 'DictionaryNotFound' occurred while
   [0] Constructing the EventProcessor
   [1] Constructing module: class=LSTInputProducer@alpaka label='hltInputLST'
   [2] Calling ProductRegistryHelper::addToRegistry, checking dictionaries for produced types
Exception Message:
No data dictionary found for the following classes:

  edm::StdArray<unsigned int,4>

Most likely each dictionary was never generated, but it may
be that it was generated in the wrong package. Please add
(or move) the specification '<class name="whatever"/>' to
the appropriate classes_def.xml file along with any other
information needed there. For example, if this class has any
transient members, you need to specify them in classes_def.xml.
Also include the class header in classes.h

A type listed above might or might not be the same as a
type declared by a producer module with the function 'produces'.
Instead it might be the type of a data member, base class,
wrapped type, or other object needed by a produced type. Below
is some additional information which lists the types declared
to be produced by a producer module that are associated with
the types whose dictionaries were not found:

  PortableHostMultiCollection<lst::HitsBaseSoALayout<128,false>,lst::PixelSeedsSoALayout<128,false> >

----- End Fatal Exception -------------------------------------------------
```
Not sure what I am doing wrong. Please let me know, if you have an idea.

VourMa commented:
Having written the previous message, I am not sure I can test anything, as I get widespread errors like:
`/data/cmsbld/jenkins/workspace/build-any-ib/w/el8_amd64_gcc12/external/alpaka/1.2.0-0ef9ed9c17707f526f6e6d8263e2898d/include/alpaka/event/EventUniformCudaHipRt.hpp(143) 'ret = TApi::eventQuery(event.getNativeHandle())' returned error  : 'cudaErrorIllegalAddress': 'an illegal memory access was encountered'!`
even when I just run the `alpaka` procModifier.

@ariostas do we have to be on the master? Could you push your latest updates and keep in sync my `VourMa/lst_inputs` branch, which is based on `CMSSW_15_1_0_pre1`?

ariostas commented:
@VourMa the `DictionaryNotFound` error is because you need to check out `DataFormats/Common` since I moved the definition of `edm::StdArray<unsigned int,4>` there since Matti said to do so.

As for your branch, I was also getting segfaults when I was trying to do things there. Even the standalone wasn't working for me. It seems like cherry-picking the commits left something inconsistent. So I think we'll need to stick with master. Maybe we can see if pre2 is ready by Monday. I can also work try generating the plots on Monday to see how they look.

VourMa commented:
> @VourMa the `DictionaryNotFound` error is because you need to check out `DataFormats/Common` since I moved the definition of `edm::StdArray<unsigned int,4>` there since Matti said to do so.

Thanks, I will try it!

> As for your branch, I was also getting segfaults when I was trying to do things there. Even the standalone wasn't working for me. It seems like cherry-picking the commits left something inconsistent. So I think we'll need to stick with master. Maybe we can see if pre2 is ready by Monday. I can also work try generating the plots on Monday to see how they look.

The segfaults I mentioned above were in this branch, not the `CMSSW_15_1_0_pre1` one. That's why I thought the other branch might work...

slava77 commented:
another resync of the target branch is in order to cleanup the history.

Also, perhaps it's time to change from Draft to ready/regular PR.
